### ISSUE TITLE
Add nn.Flatten Layer

### ISSUE THREAD

**Issue by VoVAllen** [pytorch/pytorch#2118]
This layer enables user to write conv layer and fc layer in one nn.Sequential model.

My implementation is at here:
https://gist.github.com/VoVAllen/5531c78a2d3f1ff3df772038bca37a83

Quite simple, but convenient for computer vision beginners.

**Comment by soumith** [comment]
we will not be incorporating this layer (as discussed elsewhere in some other issue / PR that i haven't found). We dont want to go down the rabbit-hole of trying to shoehorn everything into a Sequential structure, and it's an explicit choice we made. We think adding the additional three lines of adding a `forward` function is what we want to promote as good practices.

**Comment by VoVAllen** [comment]
Thanks for reply and I'm curious about the reason. Since I searched "flatten" in the issues but didn't find the related issues. Could you give more hints on how to find the related discussion?

**Comment by fmassa** [comment]
@VoVAllen I think this discussion is relevant https://discuss.pytorch.org/t/non-legacy-view-module/131

**Comment by thedch** [comment]
https://pytorch.org/docs/stable/torch.html#torch.flatten

**Comment by rahul99** [comment]
@soumith, I have a use case where I want to parse the Pytorch graph and store inbound nodes to specific layers. Since Flatten is in the `Forward` function, it will not be recorded in the graph trace. 

Specifically, I want to create a map where I can store input to specific layer indices. This will require passing input to the `torch.jit.get_trace()` This function will fail at `Linear` op due to size mismatch since the `Flatten` is defined in the forward function and not part of the trace.

**Comment by jinfagang** [comment]
@rahul99 Same problem here, do you have any solution?

**Comment by JulesGM** [comment]
I understand not wanting to have everything work in a Sequential structure, but this is a super common use case. Basically, any time you use convnets in a Sequential block, if I'm not mistaken, you would need such a layer for the final layer. This is a hugely common use-case, not some edge-case.

**Comment by black-puppydog** [comment]
[Searching for "class Flatten" "import torch.nn"](https://github.com/search?q=%22class+Flatten%22+%22import+torch.nn%22&type=Code) yields over 4000 results, most of which seem to be variations of @fmassa's [example](https://discuss.pytorch.org/t/flatten-layer-of-pytorch-build-by-sequential-container/5983/2) from the forums.

```python
class Flatten(nn.Module):
    def forward(self, input):
        return input.view(input.size(0), -1)
```

Which makes sense, since as @JulesGM points out, it is a *super* common thing to do.
*Many* modules will simply be a "hose" kind of affair: stuff goes in one end and out the other. There's no reason why this has to have more than the `forward()` method implied by a `Sequential`

**Comment by zbyte64** [comment]
[Searching for class "Lambda" "import torch.nn"](https://github.com/search?q=%22class+Lambda%22+%22import+torch.nn%22&type=Codel) yields 1400 results. This pattern allows for quick use of lambdas in defining a `Sequential` Like:

```
Lambda(lambda x: x.view(x.size(0),-1)))
```

**Comment by fmassa** [comment]
Reopening given that we reached an agreement in https://github.com/pytorch/pytorch/issues/2118 that we will be implementing this.

Thus, I believe we should add `nn.Flatten` following the semantics of `torch.flatten`.

**Comment by Chillee** [comment]
I can probably take this if nobody else is working on it.

**Comment by fmassa** [comment]
@Chillee go for it, I don't think there is anyone working on it

**Comment by Chillee** [comment]
Fixed by #22245
